### PR TITLE
DDF-2559 add ability to add custom links to the landing page

### PR DESCRIPTION
**Pull request by spearskw**
#### What does this PR do?

Adds the ability to add custom links to the landing page
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?

@AzGoalie @jckilmer
#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).

@brendan-hofmann
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).

@kcwire
@shaundmorris
#### How should this be tested? (List steps with links to updated documentation)

Add links in the landing page configuration and make sure they show up on the landing page
#### What are the relevant tickets?

[](https://codice.atlassian.net/browse/)
#### Screenshots (if appropriate)

screen shot 2016-10-27 at 10 58 08 am
screen shot 2016-10-27 at 10 58 20 am

With more prominent buttons:
screen shot 2016-10-27 at 2 44 02 pm

Bookmarks title configurable, different styling:
screen shot 2016-10-28 at 12 16 10 pm
screen shot 2016-10-28 at 12 17 05 pm
#### Checklist:
- [x] Documentation Updated
- [ ] Change Log Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: platform/landing-page/src/main/java/org/codice/ddf/landing/LandingPage.java [hunk 1]
```diff
@@ -170,6 +178,12 @@ public void setAnnouncements(List<String> announcements) {
         sortAnnouncementsByDate();
     }
 
+    public void setLinks(List<String> links) {
+        this.links = links.stream()
+                .filter(link -> !link.isEmpty())
+                .collect(Collectors.toCollection(LinkedList::new));
```
Comment (brendan-hofmann) [inline platform/landing-page/src/main/java/org/codice/ddf/landing/LandingPage.java]: Out of curiosity, why a linked list of links?

File: platform/landing-page/src/main/resources/OSGI-INF/metatype/metatype.xml [hunk 2]
```diff
@@ -55,6 +55,11 @@
                 name="Branding Logo" id="logo" required="true" type="String"
                 default=""
                 />
+        <AD
+                description="Additional links to be displayed on the landing page. Use the format <text>,<link>. For example: connexta, http://www.connexta.com"
```
Comment (brendan-hofmann) [inline platform/landing-page/src/main/resources/OSGI-INF/metatype/metatype.xml]: Please update the documentation for the landing page as well.

Comment (jlcsmith) [inline platform/landing-page/src/main/resources/OSGI-INF/metatype/metatype.xml]: Use "example.com" / example instead please

File: platform/landing-page/src/test/java/org/codice/ddf/landing/TestLandingPage.java [hunk 3]
```diff
@@ -62,6 +62,11 @@ public static void setupLandingPage() throws IOException {
                 firstDateLeadingZeroes);
         sorted = Arrays.asList(firstDateLeadingZeroes, secondDateNoLeadingZeroes, noDate);
         landingPage.setAnnouncements(unsorted);
+        List<String> links = Arrays.asList("reddit, http://www.reddit.com",
```
Comment (jlcsmith) [inline platform/landing-page/src/test/java/org/codice/ddf/landing/TestLandingPage.java]: example.com here too

File: platform/landing-page/src/main/resources/OSGI-INF/metatype/metatype.xml (hunk 2 above)
Comment (ricklarsen) [inline platform/landing-page/src/main/resources/OSGI-INF/metatype/metatype.xml]: Per @brendan-hofmann's request: send me the documentation updates as well, as the landing page PR for the draft-documentation edit has not yet merged.

**Comment by kcwire** [comment]
What does the placement look like if you have several comments? I think an administrator would want these links to be more apparent on the page.

File: platform/landing-page/src/main/java/org/codice/ddf/landing/LandingPage.java (hunk 1 above)
Comment (spearskw) [inline platform/landing-page/src/main/java/org/codice/ddf/landing/LandingPage.java]: I don't care about random access, just insertions / deletions (cause I want to ignore blank links), which a linkedlist should theoretically be better for. I can change it to whatever though. There are so few elements I doubt it matters.

**Comment by jlcsmith** [comment]
does the font seem small to anyone else? Particularly in comparison to "Additional Links"

**Comment by spearskw** [comment]
I would say the font is small, that's not just you.
I could make them more pronounced and put them at the top of the page. Maybe I could make them buttons?

File: platform/landing-page/src/test/java/org/codice/ddf/landing/TestLandingPage.java (hunk 3 above)
Comment (spearskw) [inline platform/landing-page/src/test/java/org/codice/ddf/landing/TestLandingPage.java]: Good suggestion! Dunno what I was thinking 😅

**Comment by brendan-hofmann** [comment]
Something like [bootstrap's navbar](https://getbootstrap.com/examples/navbar/), perhaps?

**Comment by michaelmenousek** [comment]
Please make the words displayed for "Additional Links" configurable so the admin could change to "Bookmarks" or whatever else.

**Comment by rzwiefel** [comment]
What!? Why make it configurable. This is one thing that absolutely has no need being configurable.

**Comment by michaelmenousek** [comment]
Wow, @rzwiefel that's very absolute.  I could see somebody wanting to call this collection of links several different things and that's why it would need to have a configurable name.

**Comment by coyotesqrl** [comment]
I still dream of the day we build with I18N and L10N in mind. So, yeah. Configurable is good.

**Comment by rzwiefel** [comment]
Okay that may have been put more harshly than needed. Of all the things we need configuration for this is probably the least necessary. Just because we *can* doesn't mean. We *should*. That's how thing become bloated. We should have our own branding, style, recognition. The more we let every joebob tinker with every part the less it becomes "us".  We should allow Integration points and branding locations... But letting them name what the 'external links' section is named is a little overboard.

**Comment by michaelmenousek** [comment]
agreed.  template languages are your friend.  jekyll, jbake, etc.

**Comment by vinamartin** [comment]
I like the navbar look 👍

File: platform/landing-page/src/test/java/org/codice/ddf/landing/TestLandingPage.java [hunk 4]
```diff
@@ -62,6 +62,12 @@ public static void setupLandingPage() throws IOException {
                 firstDateLeadingZeroes);
         sorted = Arrays.asList(firstDateLeadingZeroes, secondDateNoLeadingZeroes, noDate);
         landingPage.setAnnouncements(unsorted);
+        List<String> links = Arrays.asList("example, http://www.example.com",
```
Comment (shaundmorris) [inline platform/landing-page/src/test/java/org/codice/ddf/landing/TestLandingPage.java]: Can you please add some negative tests that handle invalid formats? Something like ",,," , ",example", "example,"

**Comment by spearskw** [comment]
Hero'd and manually tested :+1:

**Comment by shaundmorris** [comment]
Talked to @kcwire about this a little bit. Decided to merge and plan to enhance the landing page in the future.

**Pull request merged**